[refactor] Check CTxMemPool options in ctor

sedited opened the pull request:
The tests should run the same checks on the mempool options that the init code also applies. The downside to this patch is that the log line may now be printed more than once in the for loop.

This was originally noticed here https://github.com/bitcoin/bitcoin/pull/25290#discussion_r900272797.

stickies-v reviewed (commented):
Concept ACK

TheCharlatan commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -125,7 +125,9 @@ CTxMemPool MakeMempool(FuzzedDataProvider& fuzzed_data_provider, const NodeConte
-    return CTxMemPool{mempool_opts};
+    std::optional<bilingual_str> error{};
+    return CTxMemPool{mempool_opts, error};
+    Assert(!error);
```
Uff, embarassing.

stickies-v reviewed (commented):
LGTM 3aa41a31690da2ae22b6a15a9a5de0de78a01757 when fuzzer is fixed

TheCharlatan commented on src/test/fuzz/rbf.cpp:
```diff
@@ -38,7 +38,8 @@ FUZZ_TARGET(rbf, .init = initialize_rbf)
     }
 
-    CTxMemPool pool{MemPoolOptionsForTest(g_setup->m_node)};
+    auto mempool{std::move(Assert(CTxMemPool::Make(MemPoolOptionsForTest(g_setup->m_node))).value())};
+    CTxMemPool& pool = *mempool;
```
Doesn't this leave an unsafe dangling reference if we don't have a variable capturing ownership of the `unique_ptr`?

TheCharlatan commented:
Rebased 3aa41a31690da2ae22b6a15a9a5de0de78a01757 -> 571fa4f4c5c9792457a8a6ec77da980ae2d239e7 ([mempoolArgs_2](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_2) -> [mempoolArgs_3](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_3),  [compare](https://github.com/TheCharlatan/bitcoin/compare/mempoolArgs_2..mempoolArgs_3))

Updated 571fa4f4c5c9792457a8a6ec77da980ae2d239e7 -> 364456f6598b135fcc0acab8a660658b4407f837 ([mempoolArgs_3](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_3) -> [mempoolArgs_4](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_4),  [compare](https://github.com/TheCharlatan/bitcoin/compare/mempoolArgs_3..mempoolArgs_4)) 
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1422813069), using `get()` to access underlying pointer
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1422850484), limiting scope of `mempool` var in `init`.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1422862314), removed superfluous assertions.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1423915340), removed left over newline.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1423923859), renamed `Make` to `MakeUnique`.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1424003014), adjusted ownership holder variable naming to existing pattern.
* Applied @stickies-v's [patch](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1424090165) in a separate commit for fixing the fuzz test. 

I'm still not quite happy with the fuzz test, which is why I left it in a separate commit. Maybe a better approach would be to let the fuzzer do its job and just return early in the fuzz test if `MakeUnique` fails?

stickies-v commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -201,8 +201,19 @@ FUZZ_TARGET(tx_pool_standard, .init = initialize_tx_pool)
     SetMempoolConstraints(*node.args, fuzzed_data_provider);
-    CTxMemPool tx_pool_{MakeMempool(fuzzed_data_provider, node)};
-    MockedTxPool& tx_pool = *static_cast<MockedTxPool*>(&tx_pool_);
+    auto tx_pool_{MakeMempool(fuzzed_data_provider, node)};
+    while (true) {
+        if (tx_pool_) {
+            break;
+        } else {
+            // Ensure this harness is updated when additional mempool construction failure paths are introduced
+            assert(util::ErrorString(tx_pool_).original.starts_with("-maxmempool must be at least"));
+            SetMempoolConstraints(*node.args, fuzzed_data_provider);
+            tx_pool_ =  MakeMempool(fuzzed_data_provider, node);
+        }
+    }
+
+    MockedTxPool& tx_pool = *static_cast<MockedTxPool*>(tx_pool_.value().get());
```
Since we're resetting all constraints on every loop, is this not overly verbose? (+below)

<details>
<summary>git diff on 364456f659</summary>

```diff
diff --git a/src/test/fuzz/tx_pool.cpp b/src/test/fuzz/tx_pool.cpp
index 1400fadea0..7ddc45c4b4 100644
--- a/src/test/fuzz/tx_pool.cpp
+++ b/src/test/fuzz/tx_pool.cpp
@@ -200,17 +200,10 @@ FUZZ_TARGET(tx_pool_standard, .init = initialize_tx_pool)
     // The sum of the values of all spendable outpoints
     constexpr CAmount SUPPLY_TOTAL{COINBASE_MATURITY * 50 * COIN};
 
-    SetMempoolConstraints(*node.args, fuzzed_data_provider);
-    auto tx_pool_{MakeMempool(fuzzed_data_provider, node)};
-    while (true) {
-        if (tx_pool_) {
-            break;
-        } else {
-            // Ensure this harness is updated when additional mempool construction failure paths are introduced
-            assert(util::ErrorString(tx_pool_).original.starts_with("-maxmempool must be at least"));
-            SetMempoolConstraints(*node.args, fuzzed_data_provider);
-            tx_pool_ =  MakeMempool(fuzzed_data_provider, node);
-        }
+    util::Result<std::unique_ptr<CTxMemPool>> tx_pool_;
+    while (!tx_pool_) {
+        SetMempoolConstraints(*node.args, fuzzed_data_provider);
+        tx_pool_ = MakeMempool(fuzzed_data_provider, node);
     }
 
     MockedTxPool& tx_pool = *static_cast<MockedTxPool*>(tx_pool_.value().get());

```
</details>

ryanofsky commented:
This seems like an improvement, but I think the original suggestion of just adding a `std::error& error` output parameter to the CTxMemPool constructor (or a `util::Result<void>& result` output parameter) would provide a less rigid API than making the CTxMemPool constructor private and requiring every instance to be allocated on the heap and owned by a unique_ptr. This would also allow keeping the fuzz test behavior unchanged. The new fuzz test behavior is probably better, but maybe fuzz testing the max_size_bytes < descendant_limit_bytes case could find some bugs.

I don't think this pr is so high priority, so it could be closed if it too much work to update. I'm also happy to review it with or without suggested changes if it is updated.

sedited committed:
[[refactor]] Check CTxMemPool options in constructor

This ensures that the tests run the same checks on the mempool options
that the init code also applies.

TheCharlatan commented:
Reworked 364456f6598b135fcc0acab8a660658b4407f837 -> 30a1024b63105a97d04149e83ae2d8cf0830cf69 ([mempoolArgs_4](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_4) -> [mempoolArgs_5](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/mempoolArgs_4..mempoolArgs_5))

* Reworked this after @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/28830#issuecomment-2059591077). I agree that the better option here is to have an output parameter. Wasn't sure what `std::error& error` referred to, so just chose `util::Error`, but would be easy enough to use something else.

ryanofsky commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -127,7 +127,8 @@ CTxMemPool MakeMempool(FuzzedDataProvider& fuzzed_data_provider, const NodeConte
     // ...and construct a CTxMemPool from it
-    return CTxMemPool{mempool_opts};
+    util::Error error{};
+    return CTxMemPool{mempool_opts, error};
```
In commit "[[refactor]] Check CTxMemPool options in constructor" (30a1024b63105a97d04149e83ae2d8cf0830cf69)

Would be good to add comment pointing out that error is currently ignored. (And maybe that this ensures mempool code passes fuzz tests when even max mempool size is unreasonably small.)

ryanofsky reviewed (approved):
Code review ACK 30a1024b63105a97d04149e83ae2d8cf0830cf69

> Wasn't sure what `std::error& error` referred to, so just chose `util::Error`, but would be easy enough to use something else.

Sorry, that was just a confusing typo. I mean to write `std::string& error`. But I guess that doesn't work because the error message is actually translated. In that case, I do think `bilingual& error` would be better than `util::Error& error` just to make code simpler and more obvious. But no strong opinion.

TheCharlatan commented:
Thanks for taking another look at @ryanofsky!

Updated 30a1024b63105a97d04149e83ae2d8cf0830cf69 -> d447bdcfb0e38353940e4a7fc89d09482d8d39c3 ([mempoolArgs_5](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_5) -> [mempoolArgs_6](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_6), [compare](https://github.com/TheCharlatan/bitcoin/compare/mempoolArgs_5..mempoolArgs_6))

* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/28830#pullrequestreview-2022700917), replacing `util::Error` with `bilingual_str`.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1579587621), adding a comment explaining with the error in the package_eval fuzz test is ignored.

ryanofsky reviewed (approved):
Code review ACK d447bdcfb0e38353940e4a7fc89d09482d8d39c3. Thanks for the updates!

glozow commented on src/init.cpp:
```diff
@@ -1523,16 +1523,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-    }
-    LogPrintf("* Using %.1f MiB for in-memory UTXO set (plus up to %.1f MiB of unused mempool space)\n", cache_sizes.coins * (1.0 / 1024 / 1024), mempool_opts.max_size_bytes * (1.0 / 1024 / 1024));
 
     for (bool fLoaded = false; !fLoaded && !ShutdownRequested(node);) {
```
> The downside to this patch is that the log line may now be printed more than once in the for loop.

Question: do we expect this to loop many times / under what circumstances? I tried adding logging to see what happens on my node, and I haven't seen it go more than once yet.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1598392660

> Question: do we expect this to loop many times / under what circumstances? I tried adding logging to see what happens on my node, and I haven't seen it go more than once yet.

I think this is only intended to loop if there is an error loading, and the user is using a gui, and they are asked "Do you want to rebuild the block database now?" and they answer yes.

glozow replied:
Thanks @ryanofsky! That would explain why I didn't see it loop.

TheCharlatan replied:
I'm not too concerned over the message being printed again, since there a whole bunch of log lines that get printed again too. The for loop should also only be executed once more, since if there is a failure on its second pass, the `reindex` option is true, making it return an `InitError`.

glozow replied:
Thanks sgtm :+1: I didn't want to ignore the comment, but also wasn't sure how to test.

glozow reviewed (commented):
ACK d447bdcfb0e38353940e4a7fc89d09482d8d39c3 - code review, light fuzzing

TheCharlatan commented:
Rebased d447bdcfb0e38353940e4a7fc89d09482d8d39c3 -> dacdb7962c5fef8db26f6fa31facb606165d1d1e ([mempoolArgs_6](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_6) -> [mempoolArgs_7](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_7), [compare](https://github.com/TheCharlatan/bitcoin/compare/mempoolArgs_6..mempoolArgs_7))

* Fixed conflict with https://github.com/bitcoin/bitcoin/pull/29086

stickies-v commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -126,8 +126,11 @@ CTxMemPool MakeMempool(FuzzedDataProvider& fuzzed_data_provider, const NodeConte
     mempool_opts.check_ratio = 1;
     mempool_opts.require_standard = fuzzed_data_provider.ConsumeBool();
 
+    // ... ignore the error since it might be beneficial to fuzz even when the
+    // mempool size is unreasonably small
```
nit: this approach has the potential risk that if additional error handling is added to the CTxMemPool ctor, the fuzzer would silently ignore it if this test is not also updated. Could avoid that by specifically silencing only this error?

<details>
<summary>git diff on dacdb7962c</summary>

```diff
diff --git a/src/test/fuzz/package_eval.cpp b/src/test/fuzz/package_eval.cpp
index eefa566d31..1a95ff4ab6 100644
--- a/src/test/fuzz/package_eval.cpp
+++ b/src/test/fuzz/package_eval.cpp
@@ -107,7 +107,7 @@ void MockTime(FuzzedDataProvider& fuzzed_data_provider, const Chainstate& chains
     SetMockTime(time);
 }
 
-CTxMemPool MakeMempool(FuzzedDataProvider& fuzzed_data_provider, const NodeContext& node)
+std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider, const NodeContext& node)
 {
     // Take the default options for tests...
     CTxMemPool::Options mempool_opts{MemPoolOptionsForTest(node)};
@@ -126,11 +126,13 @@ CTxMemPool MakeMempool(FuzzedDataProvider& fuzzed_data_provider, const NodeConte
     mempool_opts.check_ratio = 1;
     mempool_opts.require_standard = fuzzed_data_provider.ConsumeBool();
 
-    // ... ignore the error since it might be beneficial to fuzz even when the
-    // mempool size is unreasonably small
     bilingual_str error;
     // ...and construct a CTxMemPool from it
-    return CTxMemPool{mempool_opts, error};
+    auto mempool{std::make_unique<CTxMemPool>(std::move(mempool_opts), error)};
+    // ... ignore the error since it might be beneficial to fuzz even when the
+    // mempool size is unreasonably small
+    assert(error.empty() || error.original.starts_with("-maxmempool must be at least"));
+    return mempool;
 }
 
 FUZZ_TARGET(tx_package_eval, .init = initialize_tx_pool)
@@ -152,8 +154,8 @@ FUZZ_TARGET(tx_package_eval, .init = initialize_tx_pool)
     auto outpoints_updater = std::make_shared<OutpointsUpdater>(mempool_outpoints);
     node.validation_signals->RegisterSharedValidationInterface(outpoints_updater);
 
-    CTxMemPool tx_pool_{MakeMempool(fuzzed_data_provider, node)};
-    MockedTxPool& tx_pool = *static_cast<MockedTxPool*>(&tx_pool_);
+    auto tx_pool_{MakeMempool(fuzzed_data_provider, node)};
+    MockedTxPool& tx_pool = *static_cast<MockedTxPool*>(tx_pool_.get());
 
     chainstate.SetMempool(&tx_pool);
 

```
</details>

stickies-v reviewed (approved):
ACK dacdb7962c5fef8db26f6fa31facb606165d1d1e

nit: some IWYU fixes, mostly because of the newly introduced `bilingual_str` and `Assert` usage:

<details>
<summary>git diff on dacdb7962c</summary>

```diff
diff --git a/src/test/fuzz/mini_miner.cpp b/src/test/fuzz/mini_miner.cpp
index a52c566076..3cc28a2b14 100644
--- a/src/test/fuzz/mini_miner.cpp
+++ b/src/test/fuzz/mini_miner.cpp
@@ -12,6 +12,8 @@
 #include <primitives/transaction.h>
 #include <random.h>
 #include <txmempool.h>
+#include <util/check.h>
+#include <util/translation.h>
 
 #include <deque>
 #include <vector>
diff --git a/src/test/fuzz/package_eval.cpp b/src/test/fuzz/package_eval.cpp
index eefa566d31..9b5019696e 100644
--- a/src/test/fuzz/package_eval.cpp
+++ b/src/test/fuzz/package_eval.cpp
@@ -16,6 +16,7 @@
 #include <test/util/setup_common.h>
 #include <test/util/txmempool.h>
 #include <util/rbf.h>
+#include <util/translation.h>
 #include <validation.h>
 #include <validationinterface.h>
 
diff --git a/src/test/fuzz/partially_downloaded_block.cpp b/src/test/fuzz/partially_downloaded_block.cpp
index 8a13a12069..1faf65e6ad 100644
--- a/src/test/fuzz/partially_downloaded_block.cpp
+++ b/src/test/fuzz/partially_downloaded_block.cpp
@@ -10,6 +10,9 @@
 #include <test/util/setup_common.h>
 #include <test/util/txmempool.h>
 #include <txmempool.h>
+#include <util/check.h>
+#include <util/translation.h>
+
 
 #include <cstddef>
 #include <cstdint>
diff --git a/src/test/fuzz/rbf.cpp b/src/test/fuzz/rbf.cpp
index 9cf053a290..4c7e70e3b0 100644
--- a/src/test/fuzz/rbf.cpp
+++ b/src/test/fuzz/rbf.cpp
@@ -13,6 +13,8 @@
 #include <test/util/setup_common.h>
 #include <test/util/txmempool.h>
 #include <txmempool.h>
+#include <util/check.h>
+#include <util/translation.h>
 
 #include <cstdint>
 #include <optional>
diff --git a/src/test/fuzz/tx_pool.cpp b/src/test/fuzz/tx_pool.cpp
index 77fd038a79..d62e9f02f2 100644
--- a/src/test/fuzz/tx_pool.cpp
+++ b/src/test/fuzz/tx_pool.cpp
@@ -15,7 +15,9 @@
 #include <test/util/script.h>
 #include <test/util/setup_common.h>
 #include <test/util/txmempool.h>
+#include <util/check.h>
 #include <util/rbf.h>
+#include <util/translation.h>
 #include <validation.h>
 #include <validationinterface.h>
 
diff --git a/src/test/fuzz/validation_load_mempool.cpp b/src/test/fuzz/validation_load_mempool.cpp
index 4b0aad8c47..51140ae039 100644
--- a/src/test/fuzz/validation_load_mempool.cpp
+++ b/src/test/fuzz/validation_load_mempool.cpp
@@ -13,7 +13,9 @@
 #include <test/util/setup_common.h>
 #include <test/util/txmempool.h>
 #include <txmempool.h>
+#include <util/check.h>
 #include <util/time.h>
+#include <util/translation.h>
 #include <validation.h>
 
 #include <cstdint>
diff --git a/src/test/miner_tests.cpp b/src/test/miner_tests.cpp
index fd153f8207..042a8a9ba5 100644
--- a/src/test/miner_tests.cpp
+++ b/src/test/miner_tests.cpp
@@ -14,8 +14,10 @@
 #include <test/util/txmempool.h>
 #include <txmempool.h>
 #include <uint256.h>
+#include <util/check.h>
 #include <util/strencodings.h>
 #include <util/time.h>
+#include <util/translation.h>
 #include <validation.h>
 #include <versionbits.h>
 
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index 816413dbb7..deea6ff27d 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -16,6 +16,7 @@
 #include <policy/settings.h>
 #include <random.h>
 #include <reverse_iterator.h>
+#include <tinyformat.h>
 #include <util/check.h>
 #include <util/feefrac.h>
 #include <util/moneystr.h>
@@ -26,6 +27,7 @@
 #include <util/translation.h>
 #include <validationinterface.h>
 
+#include <algorithm>
 #include <cmath>
 #include <numeric>
 #include <optional>

```
</details>

TheCharlatan commented on src/init.cpp:
```diff
@@ -1525,16 +1525,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     for (bool fLoaded = false; !fLoaded && !ShutdownRequested(node);) {
-        node.mempool = std::make_unique<CTxMemPool>(mempool_opts);
+        bilingual_str mempool_error;
+        node.mempool = std::make_unique<CTxMemPool>(mempool_opts, mempool_error);
+        if (!mempool_error.empty()) {
+            return InitError(mempool_error);
+        }
+        LogPrintf("* Using %.1f MiB for in-memory UTXO set (plus up to %.1f MiB of unused mempool space)\n", cache_sizes.coins * (1.0 / 1024 / 1024), mempool_opts.max_size_bytes * (1.0 / 1024 / 1024));
```
Not sure about this if it gets moved again when the for loop iterates.

stickies-v replied:
Oops no you're absolutely right, I didn't see the for-loop.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/28830#pullrequestreview-2057592650

> nit: some IWYU fixes, mostly because of the newly introduced bilingual_str and Assert usage:

A bunch of these were broken before this PR, so I wasn't sure if they make sense to repair here. For test files I stopped asking for repairing them when reviewing other PRs, it just feels too Sisyphean if they are not consistently checked by other reviewers anyway. I wish we'd start enforcing this properly though.

TheCharlatan commented:
Thank you for the review @stickies-v,

Updated dacdb7962c5fef8db26f6fa31facb606165d1d1e -> 27af3910e5fa9915d4913b82443db216d1b1d61b ([mempoolArgs_7](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_7) -> [mempoolArgs_8](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_8), [compare](https://github.com/TheCharlatan/bitcoin/compare/mempoolArgs_7..mempoolArgs_8))

* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1601457942), checking specific error message in fuzz test.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28830#pullrequestreview-2057592650), cleaned up some includes.

ryanofsky commented on src/txmempool.cpp:
```diff
@@ -395,9 +397,19 @@ void CTxMemPoolEntry::UpdateAncestorState(int32_t modifySize, CAmount modifyFee,
 
-CTxMemPool::CTxMemPool(const Options& opts)
-    : m_opts{opts}
+static CTxMemPool::Options SetOptionLimits(CTxMemPool::Options&& opts)
```
In commit "[[refactor]] Check CTxMemPool options in constructor" (27af3910e5fa9915d4913b82443db216d1b1d61b)

I think it would be clearer to avoid using && and std::move, since using these doesn't actually make things more efficient. In 27af3910e5fa9915d4913b82443db216d1b1d61b, one copy and one move of the option struct will happen. The copy happens when calling the CTxMemPool constructor (since the parameter type is a value parameter), the move happens when returning from SetOptionLimits.

But you could get the same result just using plain references and values:

```diff
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -397,19 +397,20 @@ void CTxMemPoolEntry::UpdateAncestorState(int32_t modifySize, CAmount modifyFee,
     assert(int(nSigOpCostWithAncestors) >= 0);
 }
 
-static CTxMemPool::Options SetOptionLimits(CTxMemPool::Options&& opts)
+//! Clamp option values and return error if options are not valid.
+static CTxMemPool::Options NormalizeOptions(CTxMemPool::Options opts, bilingual_str& error)
 {
     opts.check_ratio = std::clamp<int>(opts.check_ratio, 0, 1'000'000);
+    int64_t descendant_limit_bytes = opts.limits.descendant_size_vbytes * 40;
+    if (opts.max_size_bytes < 0 || opts.max_size_bytes < descendant_limit_bytes) {
+        error = strprintf(_("-maxmempool must be at least %d MB"), std::ceil(descendant_limit_bytes / 1'000'000.0));
+    }
     return opts;
 }
 
-CTxMemPool::CTxMemPool(Options opts, bilingual_str& error)
-    : m_opts{SetOptionLimits(std::move(opts))}
+CTxMemPool::CTxMemPool(const Options& opts, bilingual_str& error)
+    : m_opts{NormalizeOptions(opts, error)}
 {
-    int64_t descendant_limit_bytes = m_opts.limits.descendant_size_vbytes * 40;
-    if (m_opts.max_size_bytes < 0 || m_opts.max_size_bytes < descendant_limit_bytes) {
-        error = strprintf(_("-maxmempool must be at least %d MB"), std::ceil(descendant_limit_bytes / 1'000'000.0));
-    }
 }
 
 bool CTxMemPool::isSpent(const COutPoint& outpoint) const
```

This avoids need for std::move and &&, keeps the `CTxMemPool` constructor parameter unchanged, still does one copy and one move when calling and returning from NormalizeOptions. This also keeps error checking and clamping in the same function, which could be helpful if more options are added, because if clamping happens in one function and error checking happens in another function it could make the checking less flexible or require duplicate code.

TheCharlatan replied:
I was just re-using the pattern used for the `ChainstateManager` construction, but I like this more. Taking.

TheCharlatan replied:
>  Flatten avoid an unnecessary move:

What do you mean with this?

ryanofsky replied:
> > Flatten avoid an unnecessary move:

> What do you mean with this?

I was suggesting returning `Options&&` instead of `Options` to be consistent with the validation.cpp flatten code, and to avoid two moves potentially happening on return: One move on the `return opts;` statement which the compiler can't elide because `opts` is not a local variable, and another move after the function returns to initialize `m_opts` member.

In practice, however, it looks like only one move will happen, because even though the compiler can't apply NRVO when returning `opts` it _can_ use RVO to initialize `m_opts` in-place. So in practice returning either `Options` or `Options&&` should be equivalent, and I could see reasons for choosing either approach.

I think I still might opt to copy the validation.cpp flatten name and types so the pattern is the same both places, while also adding a `string& error` output parameter so error checking can happen alongside flattening. But any approach you want to take seems fine to me. And thanks for asking what was wrong with returning an `Options` value, because I was incorrectly assuming it was less efficient than it really is.

ryanofsky reviewed (approved):
Code review ACK 27af3910e5fa9915d4913b82443db216d1b1d61b

TheCharlatan commented:
Thanks for the reviews, going to take @ryanofsky's suggestion, hope it is not much work to review again

Updated 27af3910e5fa9915d4913b82443db216d1b1d61b -> 33f2a708e392edb2501f555efef34a558da9d717 ([mempoolArgs_8](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_8) -> [mempoolArgs_9](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_9), [compare](https://github.com/TheCharlatan/bitcoin/compare/mempoolArgs_8..mempoolArgs_9))
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1601650396), slightly changing the mechanics of the mempool constructor for better clarity.

ryanofsky reviewed (approved):
Code review ACK 33f2a708e392edb2501f555efef34a558da9d717. I actually didn't notice this push when I made my last two comments, so feel free to ignore the comments if you want to keep the current code.

TheCharlatan commented:
Sorry for the many pushes, but I think it's important to get this consistent.

Updated 33f2a708e392edb2501f555efef34a558da9d717 -> 856c8563ca342303a83977146df22768bb6e2c7e ([mempoolArgs_9](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_9) -> [mempoolArgs_10](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_10), [compare](https://github.com/TheCharlatan/bitcoin/compare/mempoolArgs_9..mempoolArgs_10))

* Changed the constructor to be more like the `ChainstateManager`.

ryanofsky reviewed (approved):
Code review ACK 856c8563ca342303a83977146df22768bb6e2c7e, just tweaked since last review to be more consistent with validation.cpp Flatten code

> Sorry for the many pushes, but I think it's important to get this consistent.

Thanks for the update! I agree it's much nicer to be consistent. And sorry for the detour I caused by not fully understanding the original approach.

stickies-v commented on src/txmempool.cpp:
```diff
@@ -395,8 +397,19 @@ void CTxMemPoolEntry::UpdateAncestorState(int32_t modifySize, CAmount modifyFee,
 
-CTxMemPool::CTxMemPool(const Options& opts)
-    : m_opts{opts}
+//! Clamp option values and return error if options are not valid.
```
nit: slightly confusing docstring: the `Options` ref is always returned, error is non-empty if options are not valid. Since `Flatten` is internal, this is probably something that mostly should be documented in the `CTxMemPool` ctor anyway though (highlighting that callers should check that `error.empty()`)

stickies-v reviewed (approved):
ACK 856c8563ca342303a83977146df22768bb6e2c7e

achow101 commented:
ACK 856c8563ca342303a83977146df22768bb6e2c7e

achow101 commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -126,7 +128,9 @@ CTxMemPool MakeMempool(FuzzedDataProvider& fuzzed_data_provider, const NodeConte
     // ...and construct a CTxMemPool from it
-    return CTxMemPool{mempool_opts};
+    bilingual_str error;
+    return CTxMemPool{mempool_opts, error};
+    Assert(error.empty());
```
```
../../../src/test/fuzz/tx_pool.cpp: In function ‘CTxMemPool {anonymous}::MakeMempool(FuzzedDataProvider&, const node::NodeContext&)’:
../../../src/test/fuzz/tx_pool.cpp:134:1: error: control reaches end of non-void function [-Werror=return-type]
  134 | }
      | ^
```

Also seems like the `Assert` is unreachable too.

TheCharlatan replied:
Ugh, I'll push a fix.

TheCharlatan commented:
Updated 856c8563ca342303a83977146df22768bb6e2c7e -> f2f50a4db2f61b767fd16ca2e99e136eb7dff8a6 ([mempoolArgs_10](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_10) -> [mempoolArgs_11](https://github.com/TheCharlatan/bitcoin/tree/mempoolArgs_11), [compare](https://github.com/TheCharlatan/bitcoin/compare/mempoolArgs_10..mempoolArgs_11))

* Addressed @achow101's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1605494321), fixing unreachable Assert.
* Slightly adjusted wording in docstring as per @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28830#discussion_r1603522559).

stickies-v commented:
re-ACK 09ef322acc0a88a9e119f74923399598984c68f6 . Fixed unreachable assert and updated docstring, and also added an exception for "-maxmempool must be at least " in the `tx_pool` fuzz test, which makes sense when looking at how the mempool options are constructed in `SetMempoolConstraints`.

ryanofsky reviewed (approved):
Code review ACK 09ef322acc0a88a9e119f74923399598984c68f6. Just fuzz test error checking fix and updated comment since last review

ryanofsky reviewed (approved):
[deleted comment accidentally posted to wrong pr]

achow101 commented:
ACK 09ef322acc0a88a9e119f74923399598984c68f6
